### ISSUE TITLE
migration timed out

### ISSUE THREAD

**Issue by dsukhoroslov** [hazelcast/hazelcast#7987]
HZ 3.6.1, 3.6.2.

I start 1 HZ node with several caches and load ~1000 data elements in it. The I start second node on the same machine. Nodes see each other and forms cluster. But subsequent data migration fails with message:

```
2016-04-19 19:09:39.230 [hz.TPoX.generic-operation.thread-10] WARN  com.hazelcast.cluster.impl.ClusterJoinManager - [localhost]:10000 [TPoX] [3.6.1] While waiting finalize join calls...
java.util.concurrent.TimeoutException: Call Invocation{serviceName='hz:core:clusterService', op=com.hazelcast.cluster.impl.operations.FinalizeJoinOperation{identityHash=918880958, serviceName='hz:core:clusterService', partitionId=-1, replicaIndex=0, callId=1291, invocationTime=1461082174202 (Tue Apr 19 19:09:34 MSK 2016), waitTimeout=-1, callTimeout=60000, members=MemberInfo{address=Address[localhost]:10000, liteMember=false} MemberInfo{address=Address[localhost]:10001, liteMember=false} , postJoinOp=com.hazelcast.cluster.impl.operations.PostJoinOperation{identityHash=578877427, serviceName='null', partitionId=-1, replicaIndex=0, callId=0, invocationTime=-1 (Thu Jan 01 02:59:59 MSK 1970), waitTimeout=-1, callTimeout=9223372036854775807, operations=[com.hazelcast.spi.impl.eventservice.impl.operations.PostJoinRegistrationOperation{identityHash=1572646343, serviceName='null', partitionId=-1, replicaIndex=0, callId=0, invocationTime=-1 (Thu Jan 01 02:59:59 MSK 1970), waitTimeout=-1, callTimeout=9223372036854775807}, com.hazelcast.spi.impl.proxyservice.impl.operations.PostJoinProxyOperation{identityHash=504057009, serviceName='hz:core:proxyService', partitionId=-1, replicaIndex=0, callId=0, invocationTime=-1 (Thu Jan 01 02:59:59 MSK 1970), waitTimeout=-1, callTimeout=9223372036854775807}, com.hazelcast.map.impl.operation.PostJoinMapOperation{identityHash=12938556, serviceName='hz:impl:mapService', partitionId=-1, replicaIndex=0, callId=0, invocationTime=-1 (Thu Jan 01 02:59:59 MSK 1970), waitTimeout=-1, callTimeout=9223372036854775807}]}}, partitionId=-1, replicaIndex=0, tryCount=100, tryPauseMillis=500, invokeCount=1, callTimeout=60000, target=Address[localhost]:10001, backupsExpected=0, backupsCompleted=0, connection=Connection [/127.0.0.1:10000 -> /127.0.0.1:10003], endpoint=Address[localhost]:10001, alive=true, type=MEMBER} encountered a timeout
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveApplicationResponse(InvocationFuture.java:367) ~[hazelcast-all-3.6.1.jar:3.6.1]
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveApplicationResponseOrThrowException(InvocationFuture.java:335) ~[hazelcast-all-3.6.1.jar:3.6.1]
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.get(InvocationFuture.java:223) ~[hazelcast-all-3.6.1.jar:3.6.1]
	at com.hazelcast.util.FutureUtil.executeWithDeadline(FutureUtil.java:294) [hazelcast-all-3.6.1.jar:3.6.1]
	at com.hazelcast.util.FutureUtil.waitWithDeadline(FutureUtil.java:278) [hazelcast-all-3.6.1.jar:3.6.1]
	at com.hazelcast.util.FutureUtil.waitWithDeadline(FutureUtil.java:252) [hazelcast-all-3.6.1.jar:3.6.1]
	at com.hazelcast.cluster.impl.ClusterJoinManager.startJoin(ClusterJoinManager.java:514) [hazelcast-all-3.6.1.jar:3.6.1]
	at com.hazelcast.cluster.impl.ClusterJoinManager.startJoinRequest(ClusterJoinManager.java:314) [hazelcast-all-3.6.1.jar:3.6.1]
	at com.hazelcast.cluster.impl.ClusterJoinManager.executeJoinRequest(ClusterJoinManager.java:231) [hazelcast-all-3.6.1.jar:3.6.1]
	at com.hazelcast.cluster.impl.ClusterJoinManager.handleJoinRequest(ClusterJoinManager.java:150) [hazelcast-all-3.6.1.jar:3.6.1]
	at com.hazelcast.cluster.impl.operations.JoinRequestOperation.run(JoinRequestOperation.java:40) [hazelcast-all-3.6.1.jar:3.6.1]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:172) [hazelcast-all-3.6.1.jar:3.6.1]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:393) [hazelcast-all-3.6.1.jar:3.6.1]
	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.processPacket(OperationThread.java:184) [hazelcast-all-3.6.1.jar:3.6.1]
	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.process(OperationThread.java:137) [hazelcast-all-3.6.1.jar:3.6.1]
	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.doRun(OperationThread.java:124) [hazelcast-all-3.6.1.jar:3.6.1]
	at com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.run(OperationThread.java:99) [hazelcast-all-3.6.1.jar:3.6.1]
2016-04-19 19:09:39.465 [hz.TPoX.migration] INFO  com.hazelcast.partition.InternalPartitionService - [localhost]:10000 [TPoX] [3.6.1] Re-partitioning cluster data... Migration queue size: 135
2016-04-19 19:09:39.498 [hz.TPoX.event-6] INFO  com.bagri.xdm.cache.hazelcast.impl.PopulationManagementImpl - migrationStarted; event: MigrationEvent{partitionId=0, status=STARTED, oldOwner=Member [localhost]:10000 this, newOwner=Member [localhost]:10001}; docs size: {}
2016-04-19 19:09:52.854 [cached1] INFO  com.hazelcast.partition.InternalPartitionService - [localhost]:10000 [TPoX] [3.6.1] Remaining migration tasks in queue => 134
2016-04-19 19:10:07.853 [cached1] INFO  com.hazelcast.partition.InternalPartitionService - [localhost]:10000 [TPoX] [3.6.1] Remaining migration tasks in queue => 134
2016-04-19 19:10:22.852 [cached16] INFO  com.hazelcast.partition.InternalPartitionService - [localhost]:10000 [TPoX] [3.6.1] Remaining migration tasks in queue => 134
2016-04-19 19:10:37.852 [cached25] INFO  com.hazelcast.partition.InternalPartitionService - [localhost]:10000 [TPoX] [3.6.1] Remaining migration tasks in queue => 134
2016-04-19 19:10:52.851 [cached1] INFO  com.hazelcast.partition.InternalPartitionService - [localhost]:10000 [TPoX] [3.6.1] Remaining migration tasks in queue => 134
2016-04-19 19:11:07.851 [cached12] INFO  com.hazelcast.partition.InternalPartitionService - [localhost]:10000 [TPoX] [3.6.1] Remaining migration tasks in queue => 134
2016-04-19 19:11:22.850 [cached12] INFO  com.hazelcast.partition.InternalPartitionService - 
```

I'm sure this scenario used to work properly with some version of HZ 3.5.x.

Please have a look,
Thanks, Denis.

**Comment by jerrinot** [comment]
@dsukhoroslov: can you share your configuration. thanks!

**Comment by dsukhoroslov** [comment]
Hi @jerrinot,

The configuration is quite complex and I don't see any way to simplify it for some reproducer test. Ok, my spring contexts are: https://github.com/dsukhoroslov/bagri/blob/master/bagri-server/bagri-server-hazelcast/src/main/resources/spring/cache-schema-context.xml, https://github.com/dsukhoroslov/bagri/blob/master/bagri-server/bagri-server-hazelcast/src/main/resources/spring/cache-system-context.xml. When node starts it first loads cache-system-context and forms administrative (system) cluster (it is a light member of system cluster). Then node check it's config and loads cache-schema-context for each schema it is participating. After node started population phase takes place and after that the node is ready for work.
Then I starts second node, with the same configs. It joins system cluster and then schema cluster. First node starts migration process to share populated data with new node. And fails with the error as above.
I can provide you all HZ logs from both nodes, or I can share my project's build and instructions on how to start these two nodes and reproduce the issue.

Thanks, Denis.

**Comment by dsukhoroslov** [comment]
Bingo!

I was able to make a reproduced test. Just start the test 1 time and when you'll see output "population finished, time to start second node!", then run second instance of the same test. I'm getting the timeout exception as above. It happens even in a single-cluster config. Looks like the issue is in the configured mapStore. If I remove the line where I get testCache from it's init method then the issue disappears. Please have a look.

`
public class MigrationTest {

	@Test
	//@Ignore
	public void testDataMigration() throws Exception {
	
		Config shConfig = new Config();
		shConfig.getNetworkConfig().setPortAutoIncrement(true);
		shConfig.getGroupConfig().setName("schema");
		shConfig.getGroupConfig().setPassword("schemapass");
		JoinConfig shJoin = shConfig.getNetworkConfig().getJoin();
		shJoin.getMulticastConfig().setEnabled(false);
		shJoin.getTcpIpConfig().addMember("localhost").setEnabled(true);
		shConfig.setProperty("hazelcast.logging.type", "slf4j");
		shConfig.setLiteMember(false);
		shConfig.getServicesConfig().setEnableDefaults(true);
		
		MapConfig mConfig = new MapConfig("test");
		mConfig.getMapStoreConfig().setEnabled(true);
		mConfig.getMapStoreConfig().setWriteDelaySeconds(10);
		mConfig.getMapStoreConfig().setImplementation(new TestCacheStore());
		shConfig.addMapConfig(mConfig);
		
		HazelcastInstance node1 = Hazelcast.newHazelcastInstance(shConfig);
		populateCache(node1, "test", 1000);

		System.out.println("population finished, time to start second node!");
		Thread.sleep(50000);
		
		//HazelcastInstance node2 = Hazelcast.newHazelcastInstance(shConfig);
		//populateCache(node, "test", 1000);

		//System.out.println("Second node started");
		//Thread.sleep(5000);
	}
	
	private void populateCache(HazelcastInstance node, String cName, int count) {

		IMap<Long, String> cache = node.getMap(cName);
		int idx = cache.size();
		for (long i=idx; i < idx+count; i++) {
			cache.set(i, String.valueOf(i));
		}
	}
	
	
	public class TestCacheStore implements MapStore<Integer, String>, MapLoaderLifecycleSupport {
		
		private IMap testCache;

		@Override
		public String load(Integer key) {
			System.out.println("received load request for key " + key);
			return null;
		}

		@Override
		public Map<Integer, String> loadAll(Collection<Integer> keys) {
			System.out.println("received loadAll request for keys " + keys);
			return null;
		}

		@Override
		public Iterable<Integer> loadAllKeys() {
			System.out.println("received loadAllKeys request");
			return null; //Collections.emptySet();
		}

		@Override
		public void init(HazelcastInstance hzInstance, Properties properties, String mapName) {
			System.out.println("received init request for map " + mapName);
			testCache = hzInstance.getMap("test2");
		}

		@Override
		public void destroy() {
			// TODO Auto-generated method stub
			
		}

		@Override
		public void store(Integer key, String value) {
			// TODO Auto-generated method stub
			
		}

		@Override
		public void storeAll(Map<Integer, String> map) {
			// TODO Auto-generated method stub
			
		}

		@Override
		public void delete(Integer key) {
			// TODO Auto-generated method stub
			
		}

		@Override
		public void deleteAll(Collection<Integer> keys) {
			// TODO Auto-generated method stub
			
		}
		
	}
	
}
`

**Comment by vbekiaris** [comment]
Thanks @dsukhoroslov for the reproducer. It seems that on the second node during `FinalizeJoinOperation` a `PostJoinMapOperation` executing on a generic thread faces contention trying to obtain a mutex, stack trace is:
```
ConcurrencyUtil:43 synchronized (mutex) {
MapServiceContextImpl:153 ConcurrencyUtil.getOrPutSynchronized(mapContainers, mapName, mapContainers, mapConstructor);
PostJoinMapOperation:123 mapServiceContext.getMapContainer(interceptorInfo.mapName);
```
Perhaps (not confirmed) this same mutex on `MapServiceContextImpl.mapContainers` is also obtained by partition threads, causing the contention. As soon as I have more concrete info I will update you.

**Comment by dsukhoroslov** [comment]
Thanks for the info @vbekiaris. Looking forward for updates.

**Comment by vbekiaris** [comment]
The problem is that while the map is being initialized on the second node, (a `PostJoinMapOperation` is one of the post join operations in the `FinalizeJoinOperation`), `MapLoaderLifecycleSupport.init` is invoked; `TestCacheStore.init` tries to obtain an instance of the same map being initialized, resulting in deadlock. Thus `PostJoinMapOperation` is never completed on node 2, resulting in a timeout on node 1. I am still investigating ways to allow this use case to work without deadlock.

**Comment by dsukhoroslov** [comment]
No, MapLoader tries to get another map: ` testCache = hzInstance.getMap("**test2**");`

**Comment by vbekiaris** [comment]
Thanks @dsukhoroslov for pointing this out, my oversight. Here are the mechanics for the resulting deadlock in the second node (with a little more detail):

- As part of the `FinalizeJoinOperation` being executed on a generic-operation thread, the `PostJoinMapOperation` gets the map container by using `MapServiceContextImpl.getMapContainer("test")`, therefore has acquired a lock on `MapServiceContextImpl.mapContainers` by entering the syncrhonized block in `ConcurrencyUtil.getOrPutSynchronized`. As the initialization of the `TestCacheStore implements MapLoaderLifecycleSupport ` is part of this code path, this thread will block waiting on a `DistributedObjectFuture.get()` (`ProxyRegistry:149`) to obtain the proxy for map "test2" while holding the mutex.
- Proxies for "test" and "test2" are being created concurrently by `PostJoinProxyOperation` execution on `cached` threads; e.g. thread `cached2` executes the `PostJoinProxyOperation` which initiates `ProxyRegistry.createProxy`; part of the process of creating the map proxy is `MapServiceContextImpl.getMapContainer("test")` which should enter the `synchronized` block in `ConcurrencyUtil.getOrPutSynchronized`. Same for another thread (e.g. `cached3`) which also executes a `PostJoinProxyOperation` that creates the proxy for map "test2". However none of these threads can enter the `synchronized` block as the mutex is being held by the generic-operation thread above.

So to summarize, the generic-operation thread executing `PostJoinMapOperation` holds the lock on `MapServiceContextImpl.mapContainers` and waits for `DistributedObjectFuture.get()` on the map proxy for "test2", while a `cached` thread which should create the map proxy for "test2" waits to acquire the lock on `MapServiceContextImpl.mapContainers` resulting in deadlock.

**Comment by dsukhoroslov** [comment]
Hi @vbekiaris,

Thanks for the details. I'm not an expert in HZ core, but may be it is possible to perform Map initialization (and its underlying MapStore initialization) before the `FinalizeJoinOperation` happens? 
What do you think about the issue itself? Is it a bug or it is a system behaviour which can not be changed? This makes `hazelcastInstance` parameter in the `MapLoaderLifecycleSupport.init` method useless: it can not be used to get any `DistributedObject` at this phase, right? And, I'm sure, this scenario used to work for me in some 3.5.x versions of HZ.
Please also note: may real case is more complicated. It wouldn't be a problem to not get a cache reference in the `init` method, but do it dynamically in the `loadAllKeys/loadAll` methods, I tried this workaround. But I still get this timeout exception, even when I do not use `hazelcastInstance` at all. 
What else can we do with this issue?

Thanks, Denis

**Comment by vbekiaris** [comment]
Hi @dsukhoroslov , there are two root causes for this issue:
1. On members who join the cluster, `MapLoaderLifecycleSupport.init` is invoked in the generic operation thread that executes `FinalizeJoinOperation`; this contradicts the requirement for post-join operations to be fast and lock-free (see [PostJoinAwareService](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/spi/PostJoinAwareService.java#L25)). Changing how `MapLoaderLifecycleSupport` is handled can be a major redesign task which may also involve breaking compatibility, so I think this is out of the question for 3.x series.
2. Additionally, creation of a `MapProxy` should be lightweight, however during construction it also creates its `MapContainer` which results in the deadlock above.

The PRs #8028 and #8029 (for `master` and `maintenance-3.x` respectively) address the current use case (obtain a reference to a `Map` other than the one in creation) by introducing finer-grained locking at the point of deadlock (which should be a general improvement, not just for this issue).

Finally, can you please elaborate on the above comment, you mention:
> But I still get this timeout exception, even when I do not use `hazelcastInstance` at all.

What sort of work is involved in your `init()` method? Does it involve creation of database connections or other lengthy operations?

**Comment by dsukhoroslov** [comment]
Hi @vbekiaris,

my current `init` implementation is quite simple and can not take more than 1ms, I believe:

```
	public void init(HazelcastInstance hazelcastInstance, Properties properties, String mapName) {
		logger.trace("init.enter; properties: {}", properties);
		hzInstance = hazelcastInstance;
		//popManager = (PopulationManagementImpl) hzInstance.getUserContext().get("popManager");
		schemaName = (String) properties.get(xdm_schema_name);
		//xmlCache = hzInstance.getMap("xdm-content");
		String df = properties.getProperty(xdm_schema_store_type);
		if (df != null) {
			dataFormat = df;
		}
	}
```

as you can see, I commented out the assignment of two most important components, so it takes just nothing. But I still get the same timeout exception.

**Comment by dsukhoroslov** [comment]
Just tried it with HZ 3.7-SNAPSHOT. The issue is still there. The only difference: now the length of migration queue is doubled:

```
     [java] 2016-04-27 23:30:51.964 [hz.TPoX.migration] INFO  com.hazelcast.internal.partition.impl.MigrationManager - [localhost]:10000 [TPoX] [3.7-S
NAPSHOT] Re-partitioning cluster data... Migration queue size: 271
     [java] 2016-04-27 23:30:51.972 [hz.TPoX.event-6] INFO  com.bagri.xdm.cache.hazelcast.impl.PopulationManagementImpl - migrationStarted; event: Mig
rationEvent{partitionId=0, status=STARTED, oldOwner=Member [localhost]:10000 this, newOwner=Member [localhost]:10001}; docs size: {}
     [java] 2016-04-27 23:30:53.451 [hz.TPoX.cached.thread-11] INFO  com.hazelcast.internal.partition.InternalPartitionService - [localhost]:10000 [TP
oX] [3.7-SNAPSHOT] Remaining migration tasks in queue => 272
```

**Comment by vbekiaris** [comment]
Your init indeed looks really minimal for a timeout to occur. Can you please add `-Dhazelcast.slow.operation.detector.stacktrace.logging.enabled=true` in the JVM options on Hazelcast startup and share any stack traces obtained from the `SlowOperationDetector`?
Can you also please test with the binary attached here
[hazelcast-3.7-SNAPSHOT.jar.zip](https://github.com/hazelcast/hazelcast/files/239881/hazelcast-3.7-SNAPSHOT.jar.zip), this is built from [my branch with the fix](https://github.com/vbekiaris/hazelcast/tree/master-fine-mutex-7987) (the changes are not yet merged into the repository).

**Comment by dsukhoroslov** [comment]
hazelcast-spring portion is missed. Could you build hazalcast-all-3.7-SNAPSHOT no try?

**Comment by vbekiaris** [comment]
Here you go (actually this is 3.6.3-SNAPSHOT with the fix).
[hazelcast-all-3.6.3-SNAPSHOT.jar.zip](https://github.com/hazelcast/hazelcast/files/240723/hazelcast-all-3.6.3-SNAPSHOT.jar.zip)

**Comment by dsukhoroslov** [comment]
ok, will try it now.

here is the log with slowOperationDetector enabled:

```
2016-04-28 17:28:25.620 [hz.TPoX.event-6] INFO  com.bagri.xdm.cache.hazelcast.impl.PopulationManagementImpl - migrationStarted; event: MigrationEvent{partitionId=0, status=STARTED, oldOwner=Member [localhost]:10000, newOwner=Member [localhost]:10001 this}; docs size: {}
2016-04-28 17:28:31.282 [hz.TPoX.SlowOperationDetectorThread] WARN  c.h.s.i.o.slowoperationdetector.SlowOperationDetector - [localhost]:10001 [TPoX] [3.6.2] Slow operation detected: com.hazelcast.cluster.impl.operations.FinalizeJoinOperation
com.hazelcast.util.ConcurrencyUtil.getOrPutSynchronized(ConcurrencyUtil.java:37)
	com.hazelcast.map.impl.MapServiceContextImpl.getMapContainer(MapServiceContextImpl.java:153)
	com.hazelcast.map.impl.operation.PostJoinMapOperation.run(PostJoinMapOperation.java:116)
	com.hazelcast.cluster.impl.operations.PostJoinOperation.run(PostJoinOperation.java:92)
	com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:172)
	com.hazelcast.spi.impl.operationexecutor.classic.ClassicOperationExecutor.runOnCallingThread(ClassicOperationExecutor.java:392)
	com.hazelcast.spi.impl.operationservice.impl.OperationServiceImpl.runOperationOnCallingThread(OperationServiceImpl.java:298)
	com.hazelcast.cluster.impl.operations.FinalizeJoinOperation.runPostJoinOp(FinalizeJoinOperation.java:134)
	com.hazelcast.cluster.impl.operations.FinalizeJoinOperation.run(FinalizeJoinOperation.java:99)
	com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:172)
	com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:393)
	com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.processPacket(OperationThread.java:184)
	com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.process(OperationThread.java:137)
	com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.doRun(OperationThread.java:124)
	com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.run(OperationThread.java:99)
2016-04-28 17:28:36.281 [hz.TPoX.SlowOperationDetectorThread] WARN  c.h.s.i.o.slowoperationdetector.SlowOperationDetector - [localhost]:10001 [TPoX] [3.6.2] Slow operation detected: com.hazelcast.partition.impl.MigrationOperation
com.hazelcast.util.ConcurrencyUtil.getOrPutSynchronized(ConcurrencyUtil.java:37)
	com.hazelcast.map.impl.MapServiceContextImpl.getMapContainer(MapServiceContextImpl.java:153)
	com.hazelcast.map.impl.PartitionContainer.createRecordStore(PartitionContainer.java:85)
	com.hazelcast.map.impl.PartitionContainer.access$000(PartitionContainer.java:37)
	com.hazelcast.map.impl.PartitionContainer$1.createNew(PartitionContainer.java:47)
	com.hazelcast.map.impl.PartitionContainer$1.createNew(PartitionContainer.java:43)
	com.hazelcast.util.ConcurrencyUtil.getOrPutSynchronized(ConcurrencyUtil.java:40)
	com.hazelcast.map.impl.PartitionContainer.getRecordStore(PartitionContainer.java:116)
	com.hazelcast.map.impl.MapServiceContextImpl.getRecordStore(MapServiceContextImpl.java:277)
	com.hazelcast.map.impl.operation.MapReplicationOperation.run(MapReplicationOperation.java:115)
	com.hazelcast.partition.impl.MigrationOperation.runMigrationTask(MigrationOperation.java:179)
	com.hazelcast.partition.impl.MigrationOperation.migrate(MigrationOperation.java:157)
	com.hazelcast.partition.impl.MigrationOperation.doRun(MigrationOperation.java:95)
	com.hazelcast.partition.impl.MigrationOperation.run(MigrationOperation.java:81)
	com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:172)
	com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:393)
	com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.processPacket(OperationThread.java:184)
	com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.process(OperationThread.java:137)
	com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.doRun(OperationThread.java:124)
	com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.run(OperationThread.java:99)
2016-04-28 17:28:36.285 [hz.TPoX.SlowOperationDetectorThread] WARN  c.h.s.i.o.slowoperationdetector.SlowOperationDetector - [localhost]:10001 [TPoX] [3.6.2] Slow operation detected: com.hazelcast.partition.impl.ReplicaSyncResponse
com.hazelcast.util.ConcurrencyUtil.getOrPutSynchronized(ConcurrencyUtil.java:37)
	com.hazelcast.map.impl.MapServiceContextImpl.getMapContainer(MapServiceContextImpl.java:153)
	com.hazelcast.map.impl.PartitionContainer.createRecordStore(PartitionContainer.java:85)
	com.hazelcast.map.impl.PartitionContainer.access$000(PartitionContainer.java:37)
	com.hazelcast.map.impl.PartitionContainer$1.createNew(PartitionContainer.java:47)
	com.hazelcast.map.impl.PartitionContainer$1.createNew(PartitionContainer.java:43)
	com.hazelcast.util.ConcurrencyUtil.getOrPutSynchronized(ConcurrencyUtil.java:40)
	com.hazelcast.map.impl.PartitionContainer.getRecordStore(PartitionContainer.java:116)
	com.hazelcast.map.impl.MapServiceContextImpl.getRecordStore(MapServiceContextImpl.java:277)
	com.hazelcast.map.impl.operation.MapReplicationOperation.run(MapReplicationOperation.java:115)
	com.hazelcast.partition.impl.ReplicaSyncResponse.executeTasks(ReplicaSyncResponse.java:134)
	com.hazelcast.partition.impl.ReplicaSyncResponse.run(ReplicaSyncResponse.java:74)
	com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:172)
	com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:393)
	com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.processPacket(OperationThread.java:184)
	com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.process(OperationThread.java:137)
	com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.doRun(OperationThread.java:124)
	com.hazelcast.spi.impl.operationexecutor.classic.OperationThread.run(OperationThread.java:99)
2016-04-28 17:28:36.289 [hz.TPoX.SlowOperationDetectorThread] WARN  c.h.s.i.o.slowoperationdetector.SlowOperationDetector - [localhost]:10001 [TPoX] [3.6.2] Slow operation detected: com.hazelcast.partition.impl.ReplicaSyncResponse (2 invocations)
com.hazelcast.util.ConcurrencyUtil.getOrPutSynchronized(ConcurrencyUtil.java:37)
	com.hazelcast.map.impl.MapServiceContextImpl.getMapContainer(MapServiceContextImpl.java:153)
	com.hazelcast.map.impl.PartitionContainer.createRecordStore(PartitionContainer.java:85)
	(...)
2016-04-28 17:28:36.298 [hz.TPoX.SlowOperationDetectorThread] WARN  c.h.s.i.o.slowoperationdetector.SlowOperationDetector - [localhost]:10001 [TPoX] [3.6.2] Slow operation detected: com.hazelcast.partition.impl.ReplicaSyncResponse (3 invocations)
com.hazelcast.util.ConcurrencyUtil.getOrPutSynchronized(ConcurrencyUtil.java:37)
	com.hazelcast.map.impl.MapServiceContextImpl.getMapContainer(MapServiceContextImpl.java:153)
	com.hazelcast.map.impl.PartitionContainer.createRecordStore(PartitionContainer.java:85)
	(...)
2016-04-28 17:28:36.310 [hz.TPoX.SlowOperationDetectorThread] WARN  c.h.s.i.o.slowoperationdetector.SlowOperationDetector - [localhost]:10001 [TPoX] [3.6.2] Slow operation detected: com.hazelcast.partition.impl.ReplicaSyncResponse (4 invocations)
com.hazelcast.util.ConcurrencyUtil.getOrPutSynchronized(ConcurrencyUtil.java:37)
	com.hazelcast.map.impl.MapServiceContextImpl.getMapContainer(MapServiceContextImpl.java:153)
	com.hazelcast.map.impl.PartitionContainer.createRecordStore(PartitionContainer.java:85)
	(...)
2016-04-28 17:28:37.280 [hz.TPoX.SlowOperationDetectorThread] WARN  c.h.s.i.o.slowoperationdetector.SlowOperationDetector - [localhost]:10001 [TPoX] [3.6.2] Slow operation detected: com.hazelcast.partition.impl.ReplicaSyncResponse (5 invocations)
com.hazelcast.util.ConcurrencyUtil.getOrPutSynchronized(ConcurrencyUtil.java:37)
	com.hazelcast.map.impl.MapServiceContextImpl.getMapContainer(MapServiceContextImpl.java:153)
	com.hazelcast.map.impl.PartitionContainer.createRecordStore(PartitionContainer.java:85)
	(...)
```

**Comment by dsukhoroslov** [comment]
Tried with the new build. This issue is still there.
I can not upload logs for some reason. .txt and .zip files are not accepted, don't know why

**Comment by vbekiaris** [comment]
Hi Dennis, can you verify that the `MigrationTest` you posted above works with the binary I attached?

**Comment by dsukhoroslov** [comment]
Yes, the test works properly now. But not the app :(.

**Comment by vbekiaris** [comment]
Hi Dennis,
can you please provide the slow operation stack traces obtained when running your app with hazelcast-all-3.6.3-SNAPSHOT.jar? Can I reproduce this locally using https://github.com/dsukhoroslov/bagri ?

Thanks a lot for your help with tracking this issue.

**Comment by dsukhoroslov** [comment]
Hi @vbekiaris,

Yes, you should be able to reproduce the issue. Just download bagri master codebase from GitHub. build it from root dir with maven: `> mvn clean install`. Then go to `bagri-server/bagri-server-hazelcast` dir and start first node: `> mvn -Drun-server -DskipTests -Dprofile=second -Dnode.instance=0 -Ptest package`. when node successfully starts you'll see output that schema was initialized. Then you can start second node: `> mvn -Drun-server -DskipTests -Dprofile=second -Dnode.instance=1 -Ptest package`. Here you should see the issue. 
If this does not work for you, I can provide a build, it'll be easier to start nodes with it.
I'll provide slow operation logs a bit later today, can not collect them right now.

Thanks, Denis

**Comment by vbekiaris** [comment]
Thanks Denis, I can reproduce the issue locally and I obtained the slow ops stack traces.

**Comment by dsukhoroslov** [comment]
any luck with this issue @vbekiaris? I tried to find some workaround for it. if I disable mapStore completely then it is gone. If I assign another type of mapStore (dummyMapStore, does nothing except logging) then it is also gone. But if I do register the present mapStore and do nothing in it - the timeout happens..

**Comment by vbekiaris** [comment]
Hi Denis, we have fixed certain deadlock conditions in #8105.
There is a change you will need to apply to bagri-server and, as you already noticed, relates to [MapStoreFactory](https://github.com/vbekiaris/hazelcast/blob/f6830c61ae495201ae6839154940a663f2eb32f6/hazelcast/src/main/java/com/hazelcast/core/MapStoreFactory.java#L28) implementation: `MapStoreFactory.newMapStore` should not block, as it is invoked as a post-join operation. However in `XDMMapStoreFactory.newMapStore` there is an invocation to `refresh` on a Spring context; this method may (and  actually will, in your use case) block in Spring's `AbstractApplicationEventMulticaster:178`, resulting in unavoidable deadlock.
When I switched the implementation to directly instantiate a `DocumentCacheStore` as follows, the Hazelcast cluster was properly joined by the second member:
```java
@Override
	public MapLoader newMapStore(String mapName, Properties properties) {
		String schemaName = properties.getProperty(xdm_schema_name);
		String type = properties.getProperty(xdm_schema_store_type);
		logger.debug("newMapStore.enter; got properties: {} for map: {}", properties, mapName);
		MapStore mStore = null;
		try {

			mStore = new DocumentCacheStore();
			((DocumentCacheStore)mStore).setDataPath(properties.getProperty(xdm_schema_store_data_path));
		
			if (type == null) {
				// but map-store is enabled! means: config error
				// todo: throw exception
				mStore = new DummyCacheStore();
			}
		} catch (Exception ex) {
    		logger.error("newMapStore.error: ", ex.getMessage(), ex);
		}
		logger.debug("newMapStore.exit; returning: {}", mStore);
		return mStore; 
	}
```

**Comment by dsukhoroslov** [comment]
HI @vbekiaris,

Yes, this is the case with dummyMapStore, actually: it is created directly. Ok, I'll try to get mapStore dynamically in some other way. Thank you so much!

Denis.

**Comment by dsukhoroslov** [comment]
Hi @vbekiaris,

I refactored  my MapStoreFactory, now the issue is gone with your fix for 3.6.3-SNAPSHOT. When the 3.6.3 release will be available?

Thank you so much for your help, 
Denis.

**Comment by vbekiaris** [comment]
Great news Denis, thanks for sharing. There is still a bit of work & review to be done for the fix to be merged; 3.6.3 is due soon, hope we can make it.

**Comment by vbekiaris** [comment]
Closed via #8105 , fix will be available in 3.6.3 and will be forward ported to 3.7. Thank you @dsukhoroslov for reporting this.

**Comment by dsukhoroslov** [comment]
Great! Thank you for the quick issue resolution!
